akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -318,55 +318,35 @@ private[akka] final case class Grouped[T](n: Int) extends PushPullStage[T, immut
  * INTERNAL API
  */
 private[akka] final case class Sliding[T](n: Int, step: Int) extends PushPullStage[T, immutable.Seq[T]] {
-  private val buf = {
-    val b = Vector.newBuilder[T]
-    b.sizeHint(n)
-    b
-  }
-  var bufferedElements = 0
+  private var buf = Vector.empty[T]
+  private var bufferedElements = 0
 
   override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
-    buf += elem
+    buf :+= elem
     bufferedElements += 1
     if (bufferedElements < n) {
       ctx.pull()
     } else if (bufferedElements == n) {
-      ctx.push(buf.result())
+      ctx.push(buf)
     } else {
-      if (step > n) {
-        if (bufferedElements == step) {
-          buf.clear()
-          buf.sizeHint(n)
-          bufferedElements = 0
-          ctx.pull()
-        } else {
-          ctx.pull()
-        }
-      } else {
-        val emit = buf.result()
-        buf.clear()
-        buf.sizeHint(n)
-        emit.drop(step).foreach(buf += _)
-        val updatedEmit = buf.result()
-        bufferedElements = updatedEmit.size
-        if (bufferedElements == n) ctx.push(updatedEmit)
-        else ctx.pull()
-      }
+      buf = if (step > n) {
+        if (bufferedElements == step) Vector.empty
+        else buf
+      } else buf.drop(step)
+
+      bufferedElements = buf.size
+      if (bufferedElements == n) ctx.push(buf)
+      else ctx.pull()
     }
   }
 
   override def onPull(ctx: Context[immutable.Seq[T]]): SyncDirective =
-    if (ctx.isFinishing) {
-      val emit = buf.result()
-      if (emit.size == n) {
-        ctx.finish()
-      } else {
-        ctx.pushAndFinish(emit)
-      }
-    } else ctx.pull()
+    if (!ctx.isFinishing) ctx.pull()
+    else if (buf.size == n || step > n) ctx.finish()
```

Comment:
What do you mean in terms of code? (I can try that)

-- 
Cheers,
√
On 12 Dec 2015 18:15, "Roland Kuhn" <notifications@github.com> wrote:

> In akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
> <https://github.com/akka/akka/pull/19153#discussion_r47434154>:
>
> >      }
> >    }
> >
> >    override def onPull(ctx: Context[immutable.Seq[T]]): SyncDirective =
> > -    if (ctx.isFinishing) {
> > -      val emit = buf.result()
> > -      if (emit.size == n) {
> > -        ctx.finish()
> > -      } else {
> > -        ctx.pushAndFinish(emit)
> > -      }
> > -    } else ctx.pull()
> > +    if (!ctx.isFinishing) ctx.pull()
> > +    else if (buf.size == n || step > n) ctx.finish()
>
> But this line is not free from bugs: when the buffer is not yet full
> enough, we must emit regardless of whether step > n or not.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19153/files#r47434154>.
>